File: Problem2.py
class Solution:
    def calculateTax(self, taxBracket: List[List[int]], income: int) -> float:
        '''
        Time complexity -s O(N), N --> number of tax brackets. The loop iterates through each tax bracket once.
        Space Complexity - O(1)
        '''
        # Check if the taxBracket is empty, in which case no tax is applicable
        if not taxBracket:
            return 0.0

        tax = 0.0  # Initialize the tax amount to 0
        prev_upper = 0  # Initialize the upper bound of the previous tax bracket

        # Iterate through each tax bracket to calculate the tax amount
        for upper, rate in taxBracket:
            # Calculate the taxable income within the current tax bracket
            taxable_income = min(upper, income) - prev_upper

            # If the taxable income is positive, calculate the tax amount and add it to the total
            if taxable_income > 0:
                tax += taxable_income * (rate / 100)

            prev_upper = upper  # Update the previous upper bound for the next iteration

        return tax

File: Problem2.java
class Solution {
    //TC: O(n)
  //SC: O(1)
    public double calculateTax(int[][] brackets, int income) {
        int  upper1 = 0;
        double tax = 0;
        for(int [] bracket : brackets){
            int percent = bracket[1];
            int upper2 = bracket[0];
            int taxable = Math.min(income,upper2 - upper1);
            tax += (percent * taxable)/100.0;
            income -= taxable;
            upper1 = upper2;
        }
        return tax; 
    }
}
